### ISSUE TITLE
Enable LaTeX to be added to the preamble of a tikz chunk.

### ISSUE THREAD

**Issue by krivit** [yihui/knitr#1735]
This is a proposed enhancement to the tikz engine. If there is interest, I volunteer to implement it.

Currently, there is no way to modify the preamble of an `engine="tikz"` chunk short of using a custom template via `engine.opts=list(template=...)`. I propose to add more engine options to enable the user to add extra packages or arbitrary LaTeX to the preamble. Some possible semantics:

* `engine.opts=list(extra_dependencies = c("amssymb"))` (patterned after `rmarkdown`): insert `\usepackage{amssymb}` into the preamble
* `engine.opts=list(extra_tikz_dependencies = c("trees"))`: insert `\usetikzlibrary{trees}`
* `engine.opts=list(header_includes = c("\\usepackage{amssymb}", "\\usetikzlibrary{trees}"))`: concatenate the inputs, separated by `\n` and insert into the preamble.

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.name/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('knitr')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('yihui/knitr')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by yihui** [comment]
The request sounds good to me. Thank you!

**Comment by OR-STID** [comment]
Very interesting project.
I use tikz chunk to create graphics (tikz+latex) in various formats (html,docx,pdf) and it would be great to have the possibility of adding latex packages in the preamble.  Very usefull for the projects of my students, for the package they have to build. Thank you.

**Comment by krivit** [comment]
I haven't forgotten this, but it's been a very busy term. I'll probably get a chance to implement this in December or January.

**Comment by OR-STID** [comment]
Thank you Pavel, 

we bypass the difficulty, for the moment, by adding our latex packages in the file tikz2pdf.tex of the knitr package. 
But it's a poor solution, by hand. Not to reproduce every time/ every where. Students have to produce their package in june and have a lot to learn and code before. 

I enjoy your help in advance, 
have my support, 

Olivier 
----- Mail original -----

De: "Pavel N. Krivitsky" <notifications@github.com> 
À: "yihui/knitr" <knitr@noreply.github.com> 
Cc: "OR-STID" <olivier.renault@univ-grenoble-alpes.fr>, "Manual" <manual@noreply.github.com> 
Envoyé: Dimanche 10 Novembre 2019 21:53:54 
Objet: Re: [yihui/knitr] Enable LaTeX to be added to the preamble of a tikz chunk. (#1735) 



I haven't forgotten this, but it's been a very busy term. I'll probably get a chance to implement this in December or January.

**Comment by krivit** [comment]
Apologies for taking so long to get to what is a 6-line patch (implementing the third semantics). Ideally, the engine would also copy any `header-includes:` settings, but I am not sure how to get that particular global setting from within `eng_tikz()`.

@yihui, if 1467419 is satisfactory, I'll create a PR.

**Comment by krivit** [comment]
Also, I am not sure where and how it should be documented.

**Comment by krivit** [comment]
I've stumbled on to this issue in one of my other projects, so I thought I would come back to it. @OR-STID , does PR #1886 work for you?

@yihui , apologies for bothering you, but I want to confirm if this feature still wanted and whether the interface implemented in #1886 is acceptable or whether you want me to implement one of the others?

**Comment by OR-STID** [comment]
I would like to test this new functionality. 
Is it possible to have an example of how to add latex lines to the header of the tikz chunk?.  For instance,
I want to add the following lines to my header.
\usepackage{xcolor}
\usepackage{multirow}
\usepackage{colortbl}
\usepackage{hhline}

**Comment by OR-STID** [comment]
This feature is still wanted and will be used by our next package.

**Comment by krivit** [comment]
@OR-STID , I ended up implementing the third option, because it was the most flexible. The way you would write it is therefore
```r
```{tikz, engine.opts=list(extra_preamble=c("\\usepackage{xcolor}", "\\usepackage{multirow}", "\\usepackage{colortbl}", "\\usepackage{hhline}"))}
```
followed by the TikZ code. I am still not 100% sure where to document it.

Though, on another look, putting a dot instead of the underscore in `extra_preamble` should be more consistent. @yihui ?

**Comment by OR-STID** [comment]
It's ok for me.
We can now produce output  similar to the gold standard SAS PROC TABULATE, directly in rmarkdown, without the SAS kernel. 

With the previous packages,

`\begin{tabular}{|>{\columncolor{gray!40}}l|>{\columncolor{gray!40}}l|c|c|}
\hline
\rowcolor{gray!40}\multicolumn{2}{|c|}{}&\multicolumn{2}{c|}{Product}\\
\hhline{|>{\arrayrulecolor{gray!40}}-->{\arrayrulecolor{black}}|--|}
\rowcolor{gray!40}\multicolumn{2}{|c|}{}&CHAIR&SOFA\\
\hhline{|>{\arrayrulecolor{gray!40}}-->{\arrayrulecolor{black}}|--|}
\rowcolor{gray!40}\multicolumn{2}{|c|}{}&Mean&Mean\\
\hline
Country&&&\\
\hhline{|--|~|~|}
CANADA& Actual Sales&523.32&522.24\\
\hhline{|>{\arrayrulecolor{gray!40}}->{\arrayrulecolor{black}}|-|-|-}
& Predicted Sales&487.46&476.31\\
\hline
GERMANY& Actual Sales&490.68&573.54\\
\hhline{|>{\arrayrulecolor{gray!40}}->{\arrayrulecolor{black}}|-|-|-}
& Predicted Sales&459.05&515.80\\
\hline
U.S.A& Actual Sales&530.58&452.01\\
\hhline{|>{\arrayrulecolor{gray!40}}->{\arrayrulecolor{black}}|-|-|-}
& Predicted Sales&471.30&470.92\\
\hline
\end{tabular}`

works perfectly! 
It will simplify some part of their project, programming such structures and their contents from dataframes.
Great thanks  👍

**Comment by yihui** [comment]
I just merged the PR #1886 but please note that I renamed the option name from `extra_preamble` to `extra.preamble`. Thanks!
